### ISSUE TITLE
Support for omemo

### ISSUE THREAD

**Issue by yelloff** [jitsi/jitsi#199]
That would be nice.

**Comment by ibauersachs** [comment]
What is that?

**Comment by fippo** [comment]
@ibauersachs: an e2e encryption thing:
http://xmpp.org/extensions/inbox/omemo.html -- see also some recent discussion on standards@xmpp.org

**Comment by jitsi-developers** [comment]
This is extraordinary weird. I just viewed this article 6 hours ago and
suddenly this e-mail came !

https://xmpp.org/2015/12/xmpp-at-the-end-of-the-google-summer-of-code-2015/
(2nd paragraph, Andreas Straub)

On 12/25/2015 09:57 PM, Ingo Bauersachs wrote:
>
> What is that?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jitsi/jitsi/issues/199#issuecomment-167260476>.
>
>
>
> _______________________________________________
> dev mailing list
> dev@jitsi.org
> Unsubscribe instructions and other list options:
> http://lists.jitsi.org/mailman/listinfo/dev

-- 
/Regards/,
/Ahmad/

**Comment by yelloff** [comment]
I was looking for an xmpp based alternative to Text Secure/Signal. And found this blog:
http://www.titus-stahl.de/blog/2015/09/15/how-to-get-secure-decentralized-chat-working-with-xmpp-both-on-mobile-and-desktop/

**Comment by lovetox** [comment]
this would be great, there is no desktop client that supports OMEMO
normal OTR from desktop -> phone is not the best UX.

also, all the desktop clients from whatsapp and singal need a phonenumber and a smartphone to activate an account -> horrible.

**Comment by Evi1M4chine** [comment]
I just came over from [Freedesktop’s Bugzilla for Telepathy’s XMPP component (gabble)](https://bugs.freedesktop.org/show_bug.cgi?id=93090) and while [they also have a bug for it](https://bugs.freedesktop.org/show_bug.cgi?id=93090), the [KDE developers downstream don’t expect much from them](https://bugs.kde.org/show_bug.cgi?id=355423). So Jitsi is probably our only hope for now.

(It it would enable chatting with Signal/TextSecure/Axolotl users, that would be the single greatest thing for desktop messaging as of now.)

**Comment by ibauersachs** [comment]
This would need to come from the community, we simply don't have the resources to develop this.

**Comment by lovetox** [comment]
there is work on a Gajim Plugin https://github.com/kalkin/gajim-omemo
it already works for me.

@Evi1M4chine how do you think this would enable chatting with Signal or TextSecure Apps?
even if it is the same tech behind it, the companys behind it would have to allow it, if this was possible i think many people already would have made desktop apps for it.

**Comment by improti** [comment]
The implementation contained in https://github.com/siacs/Conversations could also be portable. It's Java on Android, but I got no clue about the libraries used. Any chance to get the developers of these projects to make a common backend and provide their distinct UIs on top of that or something? There was a Jitsi on Android initiative anyways, right? The UI and design principles of Conversations make a really good impression, if you don't know it maybe take a look. It's chat-only though, i.e. without VoIP. Pictures in chats are a killer-feature, imho. Getting that and the OMEMO goodies working between the two (no matter if with common backend or not) would be really awesome.

**Comment by Evi1M4chine** [comment]
##### @lovetox:
Why would you think they couldn’t? The protocol is the same, so it’s only a matter of picking the same server to connect to. But you don’t even have to do that, since Signal’s server, being just an XMPP one with Axolotl/Omemo added in, supports federation (which is already used, e.g. with CyanogenMod’s builtin SMS/messaging app).
I also made a Signal fork this morning. Other than warnings about chat partners having different keys there were no problems. So obviously the “companies* behind it” don’t prevent it.

\___
\* Which I assume are not criminals/psychopaths and hence don’t employ nasty anti-social behavior like exclusivism or trying to corporate-murder each other.

___
##### @improti:
Signal (formerly TextSecure) itself is already Android/Java. I don’t think it’s a problem to take any necessary code from there. It’s more a problem of: Who wants to do it?

**Comment by oodbur** [comment]
OMEMO could be significant: The Axolotl ratchet is the much improved and widely used (see WhatsApp and several others) successor to OTR. Now OMEMO integrates this directly into XMPP, so no more of the criticism of operating on the wrong layer of the protocol stack, which hurdled adoption (see Empathy/Telepathy) and is connected to usability problems of OTR.

**Comment by jitsi-developers** [comment]
if OMEMO is useful/significant/superior to OTR etc etc - could you
please provide the plugin/support for Jitsi for it?

currently users (incl. myself) are happy that Jitsi client function is
maintained as it is because only very few developers are currently
volunteering to support Jitsi client

**Comment by Evi1M4chine** [comment]
Frankly, without OMEMO, there is no point in even using an IM client. OTR is unacceptable from a usability and design standpoint, and IM clients without push abilities plainly can’t work on mobile devices without dragging down the battery or not receiving messages.
So this is really the only choice. And Axolotl/OMEMO is apparently a rather nice implementation.
Which is why it is rapidly becoming the standard.
So any IM client who can’t do OMEMO, soon won’t be compatible with anything… especially not federated mobile XMPP clients.
If I can do anything about it, I won’t even allow non-OMEMO connections on my server or any client I might write or use, to show a big middle finger to any state terrorists (aka “intelligence” agencies, etc).

So, even more frankly, this is simply a question of life and death for Jitsi.
No OMEMO, and you could as well close down now, instead of letting it rot to a slow unavoidable [digital donkey](http://thedailywtf.com/articles/Virtudyne_0x3a__The_Digital_Donkey) undeath.

The only thing I am ashamed of, is that I have so little resources to give, to support your great software.

**Comment by improti** [comment]
An implementation of OMEMO for XMPP is available (in Java) in the referenced project (https://github.com/siacs/Conversations). I'd very much like to see that becoming used in Jitsi (as well as the Jitsi ZRTP over XMPP becoming used in Conversations). I cannot do that atm due to large overhead involved (I know neither of the projects code-wise). But could you please try to get in touch with them and see if you guys can help each other here? Maybe drop a note in https://github.com/siacs/Conversations/issues/1646 or something? It might actually be a rather small task if you could exchange code/knowledge here?

**Comment by turint** [comment]
Just like it was already mentioned by the devs -- yes, OMEMO is nice thing, but the code has to come from the community. Jitsi is a big project, with lot of work there to maintain it and believe me, the devs won't reject a good code attribution.

So once again, it's "talk is cheap, show me the code".

**Comment by improti** [comment]
@turing: Here is the code: https://github.com/siacs/Conversations/tree/master/src/main/java/eu/siacs/conversations/crypto/axolotl - as said, I have no clue how Jitsi is build and how to go and integrate it (as I do not know how it is used there) :( If you tell me what info you need I would actually go and try to get it for you, although I think a direct exchange between you guys would be far more effective (and could yield synergies in the future as well?).

**Comment by lovetox** [comment]
@improti i think you misunderstanding him.

its not a question of information to make this possible, like you are trying to provide.
its a question of having the time to implement it.

it seems that the developers dont have time, even if it is a great feature.
so he is asking for a code contribution, someone from outside of the current development team to make that contribution.

so if you cant code it yourself or get someone else to code it, you simply cant help.

**Comment by prinzpi** [comment]
Hi,
any new information regarding the case about the OMEMO support?

**Comment by ibauersachs** [comment]
I'm not even capable of maintaining the existing features. So, no.

**Comment by Evi1M4chine** [comment]
Maybe there needs to be a bug about creating a funding campaign? :)

**Comment by DreamFlasher** [comment]
I'd love to see OMEMO being supported by Jitsi.

**Comment by mrbiber** [comment]
[Bounty](https://www.bountysource.com/issues/29398390-support-for-omemo) at bountysource

**Comment by Ashaman-** [comment]
Added $25 to the bounty. Will add another $100 in if/when someone starts working on it.

**Comment by hoijui** [comment]
can omemo also be used for voice and video communication/streams, or is it just for text messages and file transfers?

**Comment by Ashaman-** [comment]
@hoijui Text and file transfer, with multi device and group message support. Zrtp is the standard for voice and video encryption - it's what Signal uses too.

On June 8, 2016 5:00:30 PM EDT, hoijui <notifications@github.com> wrote:
>can omemo also be used for voice and video communication/streams, or is
>it just for text messages and file transfers?
>
>---
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/jitsi/jitsi/issues/199#issuecomment-224725968

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by SamWhited** [comment]
On Wed, Jun 8, 2016 at 4:00 PM, hoijui <notifications@github.com> wrote:
> can omemo also be used for voice and video communication/streams, or is it just for text messages and file transfers?

OMEMO is a wire format designed for encrypting messages in XMPP (so
it's not well suited for video or file transfers; the current file
transfer mechanism doesn't really count since it requires receiving
the entire file before you can begin decryption, and it's actually
encrypted/decrypted out of band with a separate AES key).

ZRTP is a much better fit for A/V streams.

—Sam

**Comment by JohnnyCalavera** [comment]
Is anybody working on this? Seems like so many people are talking about OMEMO, but there is no stable desktop solution and on mobile devices there is only conversations. 
At least Chatsecure is working on OMEMO integration... Still, we need something for our desktop systems.

**Comment by DreamFlasher** [comment]
There is a stable desktop solution which works on the relevent desktop OS: [Gajim](https://gajim.org) -- and it's the reason why I switched to it, away from Jitsi. So yeah Jitsi, would be great if you add that feature, because apart from that you have a great software :)

**Comment by Evi1M4chine** [comment]
The thing is: OMEMO and the ratchet design only became necessary on unstable mobile connections. Jitsi already includes proper encryption. The only reason I would like to see it, is so it can communicate nicely with mobile messengers that need it. (Since all [my] messengers should use the same standardized protocol to avoid lock-in.)

**Comment by DreamFlasher** [comment]
The thing is: That's wrong.
See: https://conversations.im/omemo/

![comparison](https://media.kuketz.de/blog/artikel/2016/conversations/OMEMO.jpg)

**Comment by ccoenen** [comment]
@Evi1M4chine you make it sound as if OMEMO isn't all that neccessary because Jitsi is on the desktop. But you never know where the other end of your chat is received - in today's world there's a pretty high chance that at least one of the receivers is actually a mobile device with less-than-ideal connection. Also (as mentioned in the table above) receiving stuff on PC _and_ Mobile or work PC _and_ private PC (for example via XEP280 Carbon copy) is something that's fairly common. And OTR is not a good fit, there either.

I see only advantages with OMEMO, except that support is extremely limited right now. Exactly what the OP tries to get fixed.

**Comment by SamWhited** [comment]
> The thing is: OMEMO and the ratchet design only became necessary on unstable mobile connections. Jitsi already includes proper encryption.

OMEMO and the Axolotl double ratchet design have nothing to do with mobile; the main benefit OMEMO has over other end-to-end encryption mechanisms like OTR is that it can be used to send messages to multiple online clients on both ends, so your contact can get your encrypted message on their desktop and their phone. To do this it sacrifices perfect forward secrecy (it's still forward secret, but not to the same degree that OTR is), which is probably an acceptable trade off for most people. The benefit it has over something like PGP is that it is forward secret, and the down side is that it can only sync with clients that had already published pre-keys (if you add new clients, or clients don't come online for too long and the pre-keys are all used up, they won't get old messages).

It's all about the security properties you need for your use case; bandwidth and the stability of the connection isn't really a concern in any of the three major chat encryption mechanisms.

**Comment by falsefifth** [comment]
> This would need to come from the community, we simply don't have the resources to develop this.

In terms of financial units what would we be talking?

**Comment by Evi1M4chine** [comment]
I’m wondering: Isn’t the whole Signal library set already in Java, due to being Android?
And shouldn’t a transport be an independent library, used by many messengers, anyway?
So in any case, shouldn’t this be _not_ Jitsi’s job anyway?

(Also OTR misses the point, since it doesn’t do anything about the most valuable _metadata_.)

**But to bring this thing to its logical conclusion: A proper solution would be, to build full obligatory encryption of every connection right into the kernel, and add an OMEMO-equivalent as a facility to that!**
(And this encryption should _definitely_ be decentralized, without any defective-by-design fallacy like CAs, but with pluggable encryption modules.)

Maybe this ends the discussion. :)

**Comment by vanitasvitae** [comment]
I will most definitely implement Jitsi OMEMO support as part of my bachelor thesis.

**Comment by DreamFlasher** [comment]
This might be the best bachelor thesis I've ever heard of. I wish it would be encouraged by universities to make meaningful contributions with (bachelor) projects/thesis.

**Comment by ibauersachs** [comment]
@vanitasvitae Awesome! Before you start any coding, please get in touch with us so we can make sure you're going into the right direction. This involves a lot of work.

**Comment by ibauersachs** [comment]
@falsefifth I haven't investigated the actual amount of work required. Rough estimate: two months for an experienced developer, full time, maybe more. In terms of money, this obviously depends on the country, whether you're self employed, or paying a company to do it.

**Comment by lovetox** [comment]
@ibauersachs i think you greatly overestimate the work here.

you have to set and query some stuff from a pep node
write the UI for fingerprint verification (probably the most work)
write and load some stuff from the DB

im talking about message encryption, not audio/video

**Comment by ibauersachs** [comment]
@lovetox I'm just speaking out of experience as a long time Jitsi developer. Feel free to prove me wrong :-)

The first part would be to get XEP-0313 (Message Archive Management) by updating Smack, which is by itself a huge amount of work. Only then you can even start working on Omemo.

**Comment by vanitasvitae** [comment]
> The first part would be to get XEP-0313 (Message Archive Management) by updating Smack, which is by itself a huge amount of work. Only then you can even start working on Omemo.

@ibauersachs Do you have a link or something, so I can read into that? I'm still in the phase of estimating and narrowing down my targets. Also you have a mail :)

**Comment by ibauersachs** [comment]
See issue #205 for MAM. Whatever you sent me, I can't read it (I have no PGP).

**Comment by vanitasvitae** [comment]
I resent it in cleartext :)

**Comment by SamWhited** [comment]
> The first part would be to get XEP-0313 (Message Archive Management) by updating Smack

OMEMO shouldn't actually require MAM (although one of the main benefits is that it will work with MAM, and not having MAM makes chat itself less useful, but even so it's just a "nice to have'). The only hard requirement that's not already there, IIRC, will be [PEP](https://xmpp.org/extensions/xep-0163.html) support.

**Comment by vanitasvitae** [comment]
Isn't SMACK at least already implemented in SMACK? At least the Issue [here](https://issues.igniterealtime.org/browse/SMACK-251) is closed. Or do you mean that there is still some PEP related stuff missing in Jitsis code, that isn't covered already by SMACK?

**Comment by SamWhited** [comment]
> Isn't SMACK [PEP? :)] at least already implemented in SMACK?

Yup, sorry, it is; I just meant that MAM isn't a strict requirement, updating Smack is all that's needed. The only XMPP-related thing not already in Jitsi that would be required before OMEMO is possible is PEP support.

**Comment by vanitasvitae** [comment]
In the changelog of SMACK is noted, that PEP-related classes have been deprecated in favor of pubsub. Can pubsub be used as a full replacement for PEP?

**Comment by DanScharon** [comment]
PEP is a subset of pubsub, so it _should_ work

**Comment by SamWhited** [comment]
> Can pubsub be used as a full replacement for PEP?

Yes, PEP is just a limited subset of pubsub to make it easier to implement, so you should just be able to advertise PEP, make sure you're following all the rules about when to broadcast notifications, how subscriptions work, etc. and call it a day.

**Comment by vanitasvitae** [comment]
@SamWhited So technically updating SMACK is probably not required, since Pubsub is implemented in SMACK already, right?

**Comment by MathiasRenner** [comment]
For now, Gajim supports OMEMO nicely. I just published a tutorial how to use Gajim for OMEMO:

https://medium.com/@mathiasrenner/setup-whatsapp-like-chat-messaging-with-open-source-software-complete-guide-ec7adc0d3519

**Comment by lovetox** [comment]
@MathiasRenner nice that you put in so much work.

there are a few things that are a bit unclear.

1. PGP has nothing to do with OMEMO, you make it seem like its needed to get this running. OMEMO is made so your friends dont have to go through all those key creation and key adding/publishing.
2. You cant just download the gajim-omemo plugin on ubuntu and it runs out of the box, you need various dependencys see the wiki: https://dev.gajim.org/gajim/gajim-plugins/wikis/OmemoGajimPlugin
there are some distributions who have a gajim-omemo package like debian. (this would run out of the box)

**Comment by MathiasRenner** [comment]
@lovetox  thank you for this hint! I was positively suprised to see that OpenPGP is not required. I was used to require OpenPGP for OTR :) I just updated the post.

I reply on your two other comments:
> You cant just download the gajim-omemo plugin on ubuntu and it runs out of the box, you need various dependencys see the wiki: https://dev.gajim.org/gajim/gajim-plugins/wikis/OmemoGajimPlugin
there are some distributions who have a gajim-omemo package like debian. (this would run out of the box)

See the section "2. Setup messenger on your Desktop". At the beginning, all dependencies are installed.

> To see all menus on ubuntu start type in the console export UBUNTU_MENUPROXY=0 before you start gajim

Good to know! This puts the main menu inside Gajim's window. That's a minor improvement with lower prio. 

p.s. Yes, it was and is still lots of work. But to me, this is the first moment in time that a messenger with open source, high security, based on decentralization and good usability is ready for the masses. And since I know about this and have some influence due to my IT skills, I feel responsible to spread the word. BTW, Snowden argued exactly the same :)

**Comment by DanScharon** [comment]
> @lovetox thank you for this hint! I was positively suprised to see that OpenPGP is not required. I was used to require OpenPGP for OTR :) 

You also didn't need OpenPGP for OTR. You only needed OpenPGP for OpenPGP...

> But to me, this is the first moment in time that a messenger with open source, high security, based on decentralization and good usability is ready for the masses.

with that many manual steps, it's still not ready for the masses. OMEMO should be built in and activated by default, everywhere.

**Comment by Evi1M4chine** [comment]
I don’t understand this arguing that something is only ready, if all that’s left to do, is to drool or a big red button on a touch screen, that says ”WANT”, even if that comes at the cost of pretty much every freedom of choice and everything but the dumbest common denominator of features. (Aka Apple.)

Why is it not, that people are expected to start turning on their brains if they want something, and if they don’t, then it’s their problem and natural selection will deal with it. Why do we care if the dumb masses do not _want_ to do anything to not be livestock?  Why do we give away everything that’s great, and ruin our software, just to be up their behinds like there’s no tomorrow? If they fight so hard to not have to do anything, even if that means their lives gonna be nightmares, that they yell at us if we don’t pre-chew their food _and_ drop it in their mouths, _and_ move their throat to swallow, then who are we to fight them? All they do is harm everyone and feel entitled.

We’re just as guilty for this horrible world, because we actively fight natural selection, by not letting them go under.
</rant>
(Not that that was any more off-topic than the last couple of comments. But at least it was not just “designing perfect oil lamps”.)

Now let’s stop being so OT, okay?

**Comment by DanScharon** [comment]
> Why is it not, that people are expected to start turning on their brains if they want something, and if they don’t, then it’s their problem and natural selection will deal with it. 

because we are not social darwinists. At least I'm not. And I know that a highly complex, human-made society is anything but "natural" (and hasn't been for millenia).

Maybe you should never be able to order food, since you don't know how to grow food, process food, hunt food, etc. from the bottom up. Maybe you should die out because you're so dependent on supermarkets, ovens and other devices who are making you unfit within the competition of natural selection of animals within the jungle.
Oh wait. We live in societies. These are human-made and we depend on each other in one way or another. Most of us are specialists. You for example may know about tech, but your ramblings about society sound like they come from a teenager who thinks Ayn Rand is the latest rage.

We have a responsibility for people who are not tech-savvy enough, but who would like to use the "right" tools, that we help them. If we don't care, noone will, and your ubermenschen will be drawn into the abyss with the rest of everyone else, just as "naturally" strong as they may seem.

> Now let’s stop being so OT, okay?

you let that genie out of its bottle. Don't expect people to ignore that. But yeah, you're right, it *is* OT, so I will stop with it.

**Comment by ccoenen** [comment]
It's really simple: People have lives to live. And they involve different things than fiddling with their computers and/or devices.

I'm sure someone somewhere thinks "everyone needs to know details of taxation law". No. Not everybody does. Someone surely thinks "it is imperative that people care about where and how their food is grown" - well sure that would be nice, but people don't care about that even though they literally *put that stuff into their bodies*.

So, whenever somebody cries "people should care about ..." - just fill the blank with any complicated topic that you know **nothing** about. And then imagine how much time it would take you gain a working knowledge about it. It is very convenient to cry this battlecry when it's a topic you're already familiar with. Expecting an average person with a full time job, perhaps a family and possibly some hobbies to ALSO take up your particular topic of interest is just unrealistic.

I don't get it, @Evi1M4chine, you yourself said that encryption should be at the very basis of your operating system so that everything is encrypted by default. That more or less implies that it's out of sight. Just as much out of sight as the abstraction of my networking stack (who here known their current IP without looking?) or how my SSD-Drive allocates blocks (I honestly have NO FUCKING CLUE AND I DO NOT CARE). This stuff is complicated. It needs to be abstracted away.

If we want people to use encryption, they need to have encryption by default. We do not get any regular person to encrypt their stuff if we make them jump through any single additional hoop. Not one.

**Comment by Evi1M4chine** [comment]
I just mentioned how you’re being off-topic for many comments, but if you want to blame it on me, then “[\*dons gangsta clothes\*] Hello genie! Where’s my 3 wishes, mofo??” ;))

> anything but "natural" (and hasn't been for millenia).

So the usual “nothing changes because we change nothing, and hence we’ll change nothing” excuse.
That’s what I meant with “Because we don’t _want_ to change it,”

> Maybe you should never be able to `$X`, since you don't know how to `$Y`.

Oh man, you’re saying this to the wrongest of all people. ^^
Maybe you shouldn’t assume that I can’t do things just because you and the humanoids can’t. My whole life was a journey to being able to do everything from scratch. There’s nothing between the dirt and nature outside and everything important in our society that I don’t know how to do.
But I agree with such degenerates dying out. If society wouldn’t constantly prop them up and fix their failures, most “humans” would be dead it a week. But it’s already starting to fail, as there’s nobody left to actually do it. Too bad it takes so long.

>  …and devices who are making you unfit within the competition …
> … we depend on each other …

So _that_’s the straw-men you’re using…
I didn’t say any of those things, nor imply them. So you can put those right back in the bottle where you ”found” them.

I’m not talking about useful things making one unfit. (Seriously, how the hell do you even manage to do these mental contortions? How hard is it triggering you?) I’m talking about harmful laziness.
And I’m not talking about dependency. I’m talking about choosing _what_ to enable and what not to. (The huuuge taboo of _making your actually own decisions_.)
Like not enabling harmful behavior. Which is usually the ridiculous amount of self-entitled laziness. (On a `[cost, benefit]` chart, wastefulness would be `[+1, −1]`, perfect efficiency `[±0, +1]`, and laziness `[−1, −1]`. Saving so much you end up harming yourself again.)

This is a good comparison: There’s three choices on how to do things:
1. Like Emacs~~OS~~: Massive power, all features that are thinkable, but a huge behemoth in terms of complexity. (Okay, there’s four, The HTML5 ~~operating systems~~, err I mean ”browsers”: Huge behemoths and _still_ less and _less_ power and usefulness with each release. lol)
2. Like Windows Notepad (or anything Apple): Extreme simplicity, but practically useless, unless you have no independent thought, just go with the masses and are satisfied with the lowest common denominator.
3. Like Godot Engine, or Unix-y ”everything is a file”, or Maya to some extent; and generally the whole thing that made computers special as devices: Nearly unlimited power and freedom, yet tiny complexity. Aka ”emergence”. **Because, and that’s the key point: There is zero reason to compromise on features just to remove complexity.** Either-OR compromises are for losers whose minds are too small to come up with an inclusive this-AND-that. (Not talking about Boolean operations.)

> We have a responsibility for people who are not tech-savvy enough,

Exactly! And by unnecessarily taking away their power under the excuse of simplicity, because we’re too stupid to achieve a high emergence, and thereby then enabling their laziness, by letting them get used to it, even expect it, and have the audacity to start complaining if we don’t cripple them and put them in fake-golden cages, like they’re whining spoilt brats, *we utterly fail to fulfill our responsibility.*

[P.S.: It’s ”no one”. ”Noone” would be the nickname of Data’s creator. :P]

**Comment by ccoenen** [comment]
@Evi1M4chine you completely miss the point that it is not all about you. And Data's creator was Noonien Soong.

**Comment by Evi1M4chine** [comment]
@ccoenen: You misunderstood me: Of course – I agree there – we should remove as much ”having to deal with the unnecessary” as possible. But, and that’s my point, which everyone here conveniently entirely misses, it is wrong that that would have to come at any cost to power and freedom.
Look at Godot Engine. The bigger power makes it _easier_ to use. Not harder. Making the argument about simplicity entirely invalid.

@ccoenen (your second comment): Nice! If you fail to understand the second time, you just mirror it back and add a straw-man argument.

**Comment by ccoenen** [comment]
I wish the best of luck to @vanitasvitae, @lovetox and whoever else actually works on getting this show on the road. I'll show myself out.

**Comment by ibauersachs** [comment]
Please, stop this OT discussion or I'm going to lock the ticket. For now, be happy that @vanitasvitae is probably going to work on it as part of his Bachelor's Thesis and that the protocol behind Axolotl/Signal/OMEMO is going to be documented, thus allowing non-GPL implementations.

**Comment by SamWhited** [comment]
For the record (@vanitasvitae or whomever is looking into this), OMEMO is now published as [XEP-0384](https://xmpp.org/extensions/xep-0384.html).

**Comment by vanitasvitae** [comment]
@SamWhited I'm sure, this has been asked before, but do you know if Olm and libsignal-protocol-java will be interoperable? Right now, there is no olm library for java available (if it is, please let me know!), so i'm currently implementing OMEMO using the signal lib.

**Comment by lovetox** [comment]
no they are not interoperable.
but if you have a working implementation with libsignal, its not hard to switch to olm.

the work is in the UI, Trust Management, XMPP stuff.
Not which lib you call to decrypt the message.

**Comment by DanScharon** [comment]
> Edit2: Btw: I'm implementing OMEMO as a SMACK extension.
@vanitasvitae great, based on which Smack release?

**Comment by vanitasvitae** [comment]
@DanScharon Right now I'm working with SMACK 4.2.0-beta-3 (commit d976434bb3a0f1a7107e04a34b863a198a731590).

**Comment by vanitasvitae** [comment]
Update: I'm making progress. My module is usable in 1:1 chats already. I'll tackle MUC support soon.

Btw: Take a look at this overview of [OMEMO](https://omemo.top) implementations.

**Comment by 2Belette** [comment]
@vanitasvitae your link seems broken ? 
Is there a way to test your module somewhere with Jitsi?
Many thanks ;)

**Comment by vanitasvitae** [comment]
@2Belette Yes, I separated the repository into two modules. You can find the source code here:
https://github.com/vanitasvitae/smack-omemo
https://github.com/vanitasvitae/smack-omemo-signal

**Comment by 2Belette** [comment]
Thanks !
Is it usable with Jitsi or do you plan to make it a module or integrate with the app directly?


On February 23, 2017 11:47:29 PM GMT+01:00, vanitasvitae <notifications@github.com> wrote:
>@2Belette Yes, I separated the repository into two modules. You can
>find the source code here:
>https://github.com/vanitasvitae/smack-omemo
>https://github.com/vanitasvitae/smack-omemo-signal

-- 
Sent from my mobile device. Please excuse my brevity.

**Comment by vanitasvitae** [comment]
@2Belette For now I'm focusing on making it a good usable Smack module. My goal is, that it's straight forward for a client like jitsi to implement OMEMO (sendOmemoMessage instead of sendMessage...). The hardest part on the client side should be designing the UI for activating OMEMO, trusting keys etc.

**Comment by 2Belette** [comment]
I am happy to see that you are making good improvement :)
So I probably have to wait a little bit more before someone would work to integrate your work into Jitsi UI..

Anyway you have one more supporter !



On February 24, 2017 9:41:23 AM GMT+01:00, vanitasvitae <notifications@github.com> wrote:
>@2Belette For now I'm focusing on making it a good usable Smack module.
>My goal is, that it's straight forward for a client like jitsi to
>implement OMEMO (sendOmemoMessage instead of sendMessage...). The
>hardest part on the client side should be designing the UI for
>activating OMEMO, trusting keys etc.

-- 
Sent from my mobile device. Please excuse my brevity.

**Comment by cmeng-git** [comment]
aTalk-android has implemented OMEMO. For anybody who are interested, may download from the site below. Note aTalk supports android device only:

http://atalk.sytes.net/releases/atalk-android/aTalk-release.apk

**Comment by DreamFlasher** [comment]
@cmeng-git This sounds great, but could you add a little bit more information? I can't find aTalk in the PlayStore nor on F-Droid. Also I can't find any website, http://atalk.sytes.net/ is a newly configured apache, http://atalk.sytes.net/releases/atalk-android is forbidden. Do you have the sourcecode of the apk somewhere published?

**Comment by cmeng-git** [comment]
The apk is not availabled on playstore or any of the public domain. Http://atalk.sytes.net is my home server and the cert is self signed. I have yet to release the source to the public. 

On 12 July 2017 16:48:25 GMT+08:00, Marcel Ackermann <notifications@github.com> wrote:
>@cmeng-git This sounds great, but could you add a little bit more
>information? I can't find aTalk in the PlayStore nor on F-Droid. Also I
>can't find any website, http://atalk.sytes.net/ is a newly configured
>apache, http://atalk.sytes.net/releases/atalk-android is forbidden. Do
>you have the sourcecode of the apk somewhere published?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/jitsi/jitsi/issues/199#issuecomment-314697988

--
Send from Android using XryptoMail

**Comment by DreamFlasher** [comment]
Okay, can you ping us again when you have released the source code? Before that I'd rather not install a random apk, which in the best case just shows an encryption symbol and in the worst case contains a rootkit ;)

**Comment by cmeng-git** [comment]
Understand your concern.

https://community.igniterealtime.org/people/cmeng

On 12 July 2017 19:22:18 GMT+08:00, Marcel Ackermann <notifications@github.com> wrote:
>Okay, can you ping us again when you have released the source code?
>Before that I'd rather not install a random apk, which in the best case
>just shows an encryption symbol and in the worst case contains a
>rootkit ;)
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/jitsi/jitsi/issues/199#issuecomment-314740284

--
Send from Android using XryptoMail

**Comment by louy2** [comment]
@cmeng-git In which case, are you using Smack in aTalk? Since that recently added support to OMEMO.

https://community.igniterealtime.org/blogs/ignite/2017/06/06/smack-v42-introduces-omemo-support

**Comment by cmeng-git** [comment]
Hi Louy2.

Below is the history log for aTalk that starts using Smack release version 4.2.1

===================================================================
Version:		0.8.3
Author:         cmeng
Upload Date:	07/12/2017
- Add option for OMEMO private key regeneration
- Add option to allow changing trust state for Omemo device fingerprints - stop untrusted buddies from receiving OMEMO messages
- Add option to copy crypto privateKeys or fingerprints to clipboard to assist remote verification
- Consolidate all crypto keys and fingerprints for OTR and OMEMO for better UI access
- Display system message when OMEMO chatSession is setup with unverified or untrusted buddies
- fix implementation errors for trustOmemoIdentity and distrustOmemoIdentity method - trust state unchanged
- fix no activities error due to classes relocation for OtrFingerPrints and OtrPrivateKeys (class later removed)

===================================================================
Version:		0.8.2
Author:         cmeng
Upload Date:	07/09/2017
- Implement OMEMO chat support using SQLite database for backend persistent storage
- Add user option to blind trust contact omemo key before verification on first use
- Add user option to regenerate omemo devices
- Add user option to purge unused omemo devices
- Add user option for XEP-0384 OMEMO database clean up (use with care)
- Add new background colour for OMEMO mode
- Optimize background colour changes handler and consolidate in cryptoFragment
- Enable all listeners for chatFragment only if it is primarySelected
- Update omemo database structure i.e.
  * Add omemo_devices table
  * Remove accountUuid column in all omemo table - not optimize to support multiple accounts
- Move setupOmemoConfigStore() to CryptoActivator class, needs init once only per device
- Remove OmemoManager.addOmemoMucMessageListener(this) in ChatRoomJabberImpl as OMEMO message is being handled in SmackMessageListener#processMessage

===================================================================
Version:		0.8.1
Author:         cmeng
Upload Date:	07/01/2017
- Implement OMEMO chat support using file-based persistent storage
- Fix corrected message received handler for proper edited message display

==================================================================
Version:		0.8.0
Author:         cmeng
Upload Date:	06/12/2017
- Remove lib org.whispersystems:axolotl-android:1.4.3
- Remove crypto/axolotl sources
- Remove all SQLiteAxolotlStore table in backend database
- bound 0794 to version 0.8.0 as official release

version code 0794 check in on 2017/06/12
==================================================================
Version:		0.7.9
Author:         cmeng
Upload Date:	06/12/2017
- Fixed Authorization Request prompt to show correct requester ID.
  Use contact.getAddress() for Authorization Request Dialog prompt - getDisplayName is null
- Use Smack 4.2.1-SNAPSHOT library for OMEMO support (June 2017)
- Use gmsPlayServicesVersion 11.0.0
- Need to pass db to initMCLDataBase on newly installed apk
- Block mcGroup and metaContact creation for non-persistent entity in MCIStorageManager. Non-persistent
  entity creation request when unsolicited announcement etc messages received from server e.g. atalk.org
- Need to generate non-persistent entity sessionUuid for proper message retrieve
- Change childContacts details implementation to JSONObject
- Add user option to support change contact displayName in contactList fragment (RFC 6121)
- Implement protoContactRenamed method in MCIStorageManager.class
- MessageSourceService#handleProviderAdded must have both pps registered and connected prior to calling
  Block premature called from MessageHistoryServiceImpl#handleProviderAdded
- Removal contact (waiting for approval) error - Smack 4.2.0 sends 'ask' attribute in Roster.removeEntry not allowed by RFC6121;
  fixed by Smack 4.2.1-SNAPSHOT (June 5, 2017)
- Cleanup and change OperationSetBasicInstantMessagingJabberImpl#enableDisableCarbon() to use CarbonManager.setCarbonsEnabled()
- Execute ProtocolProviderServiceJabberImpl#initRosterStore only after account has been authenticated
- Fix VCardAvatarManager#downloadVCard(BareJid userId) uncaught exception when mAccount == null. Causing avatar loading failed.

**Comment by ibauersachs** [comment]
@cmeng-git I've removed the release notes. You can link to them on you own site, provided that you release the source code. I've previously politely asked you to do that. It isn't just the morally right thing to do. Since you're now using OMEMO, thanks to the GPL, you're legally obligated to release it.

**Comment by cmeng-git** [comment]
Sorry if this has offended you. I am just reply to louy2 question. As I have mentioned earlier, atalk source will be open source. It is just that I feel the source is just no clean up enough to be released.

**Comment by ccoenen** [comment]
@ibauersachs AFAIK, it is perfectly fine to hand out source code "on request" of a user. This is not the way the community prefers it, but it is legal use under the GPL.
> ... to make sure that you have the freedom to distribute copies of free software (and charge for them if you wish), that you receive source code **or can get it if you want it** ...
*(From GPL-v3 Preamble)*

**Comment by DreamFlasher** [comment]
@cmeng-git @ccoenen Well then, I hereby request the source code of aTalk :) You can reach me via the methods listed here: https://keybase.io/dreamflasher

**Comment by cmeng-git** [comment]
Thanks for showing interest on atalk source. When I mention the source is not clean enough for public release. It is more than just pure source coding. Atalk is a fork from jitsi-android. At the moment, the source has multiple copyright statement. Leagally  I do not have good idea what it means to release source of this nature. Actually I  am about to start discussion this with someone and seek their advice. Hopefully it is not too complicated and restriction to release the source. Meantime I do not feel comfortable to release code to the public. Thank for your understanding.

**Comment by ccoenen** [comment]
@cmeng-git `jitsi-android` is [released under apache license](https://github.com/jitsi/jitsi-android/blob/master/LICENSE) at the top of this license file is a short summary what that means.

**Comment by ccoenen** [comment]
We're going off-topic a lot. I will not comment further on this branch of the discussion. Let's get back to **omemo in jitsi**.

**Comment by cmeng-git** [comment]
Hi,

 

Actually  the answer was already given in the attached history log, but was removed by the Jitsi moderator.

aTalk version 0.7.9 released on 12 June, 2017 has imported smack version 4.2.1 SNAPSHOT shortly after it was released, as it is my plan to implement OMEMO for aTalk.

 

If you want to know more, perhaps we can discuss the topic in our private mail exchange. I really do not wish to offend anybody.

 

Regards,

CM Eng

**Comment by smith476** [comment]
sorry for re-opening this thread - but client "aTalk" (Jitsi-Android fork) is now available on Google Play "aTalk (Jabber / XMPP)" which includes both OMEMO (!) and OTR support, the source is published on GitHub.
How is the progress implementing OMEMO in Jitsi-Desktop?

**Comment by DreamFlasher** [comment]
Just as a reference I found this as a good explanation about the licensing situation: https://github.com/igniterealtime/Smack/wiki/OMEMO-libsignal-Licensing-Situation
Is there any way out of that? Could there be a plugin under GPL license which wraps smack and smack-omemo-signal and then Jitsi could just use the plugin?

**Comment by vanitasvitae** [comment]
@DreamFlasher I'm not an expert on the field of software licenses. Is it possible to circumvent licensing issues with plugins?

**Comment by ibauersachs** [comment]
I think this should work. We (the Jitsi authors) don‘t mind having our code combined with GPL‘ed code, but:
- it cannot live in the Jitsi repository
- we cannot ship/distribute a combination

(because that would automatically apply the GPL to everything).

If somebody else does that though, fine. The Apache License 2.0 is compatible with the GPL3 (but not GPL2!).

**Comment by smith476** [comment]
short info: with the recent update (2.0.0)  the "Conversations" client on Android is removing support for OTR.
OMEMO is now the default encryption method.
I guess all Jitsi Desktop users would be very grateful, if their preferred client would receive OMEMO support soon, the OMEMO development with other clients is progressing:
https://omemo.top/
(I am only politely motivating and cannot provide support myself due to lack of IT skills)

**Comment by cmeng-git** [comment]
aTalk version 1.2.0 release

Adding Contact
In order to protect the privacy of XMPP users, presence information is
disclosed only to other entities that a user has approved. When a user has
agreed that another entity is allowed to view its presence, the entity is
said to have a "subscription" to the user's presence. An entity that has a
subscription to a user's presence or to which a user has a presence
subscription is called a "contact".
When you add a buddy to your contact list, aTalk actually makes a
subscription request to your buddy for his/her approval. In this process,
your hosted server routes your subscription request locally if the buddy is
on same server; otherwise remotely with connection made to the buddy hosted
server via IP derived from your buddy account service name i.e. example.org
from "buddy@example.org". If the Fully Qualified Domain Name (DQDN) of your
buddy hosted server cannot be derived from his JID account service name,
the subscription request process will fail. With the current XMPP standard
defined for Presence Subscription, it is therefore not possible to include
buddy whom resides on a different server where his/her account JID does not
contain the FQDN. You are also unable to send message to your buddy in this
case since the message is delivered via server-to-server network connection.
When your buddy client receives the Presence Subscription Request from
hosted server, an *Authorization Reqeust* dialog is launched for hist/her
approval. Your buddy may Ignore, Deny or Authorize the request.
Recipocally, a Presence Subscription Request will be reverted for your
approval when your buddy approved the reqeust. In the event your buddy
chooses to ignore your request; any further re-authorization request from
you will not trigger another Authorization Request. However the server will
resend the reqeust on your behalf when the buddy has exited and login
again. This process continues until user has made a decision either to Deny
or Authroze the request.
The "Add Contact" dialog provides the option to categorize your new contact
to one of your existing defined groups or a newly created group. You may
peform this grouping or moving to another group after the contact has been
added. The same opton is also available to your buddy when he approves the
Presence Subscription Request.
aTalk has an user selectable option for presence subscribe approval mode
i.e. manual or auto accept-all; The earlier description is when manual
approval is selected. aTalk default is auto accpet-all. Under this mode,
atalk will accept all incoming presence subscription request, and in return
rquest the same for the sender approval. The newely added contact is
grouped under Contacts.

**Comment by cmeng-git** [comment]
Can you try out and let me if found any problem?

There are still problem exist in ejabberd server and smack library. If you
see problem or no getting approval from the contact,
just logout and login the contact. - problem is at time ejabberd does not
relay the request to your contact until he logout and login again.

https://github.com/processone/ejabberd/issues/2352#issuecomment-377146175
https://discourse.igniterealtime.org/t/smack-4-2-3-unexplained-exceptions-thrown-by-smack-on-various-conditions/81235
https://discourse.igniterealtime.org/t/smack-4-2-3-throws-smackexception-noresponseexception-even-when-the-stanza-is-received-within-the-timeout-period/81112/2

**Comment by stevenroose** [comment]
@vanitasvitae How is the progress going??

**Comment by vanitasvitae** [comment]
Unfortunately I'm currently very busy with other things ([this years Summer of Code project](https://storage.googleapis.com/summerofcode-prod.appspot.com/gsoc/core_project/doc/5052450072428544_1522158049_GSoC_Proposal.pdf?Expires=1527085995&GoogleAccessId=summerofcode-prod%40appspot.gserviceaccount.com&Signature=CgKW%2F3BMEnCUkyI2i08hjwPIfRvqFNZkiXPRXinu%2FpsT8acRAjcDDfN%2Fc3xct46m%2FRtOPSoCA8c6DJ9za8zaYPvvvMksmm5gxvGrEQQeOZanSXvTLeYY17oDloS7xLRJcjRtbcF3mcETQuY1y9zoGErhF7DetDqmjaRfPnup5eCrEqtqAkmOlrtzbsCIIIqzv2bOVmfVMYGJKCiAXEC%2FZHCuXcuOoqSPab6haUnY2eJmqTzxiS3N8Pk3kHImXocenTuFwccFibozvhNQ9MstBERt7GjKws%2FcJLDgUVkp%2FmWRQaazS3z%2FZTTB86XRDBL4TLjcL1yglLsLAudUFX8zyg%3D%3D)), so I don't have much time to look at the OMEMO stuff. There are multiple heavy weight PRs I openend, which still need to get merged (the OMEMO PR is among them). However, I encourage the use of the OMEMO code in order to find bugs or - even better - submit fixes ;)

**Comment by stevenroose** [comment]
@vanitasvitae Doing OpenPGP in favor of OMEMO? :(

**Comment by vanitasvitae** [comment]
In addition :)

**Comment by MathiasRenner** [comment]
I meanwhile changed to Matrix (https://matrix.org) with the Riot client. Quite happy so far.

**Comment by ccoenen** [comment]
(good for you, not that anyone cares? If a moderator could please remove that post and mine along with it, that'd be great.)

**Comment by DreamFlasher** [comment]
It's off-topic.

**Comment by Neustradamus** [comment]
Any news?

**Comment by mario26** [comment]
I'm waiting for OMEMO in Jitsi Desktop and Jitsi Android for XMPP impatiently, could it work soon? 

Thank you !

**Comment by vanitasvitae** [comment]
I'm sure aTalk can need some contributors ;)

**Comment by cmeng-git** [comment]
aTalk is working well now even for Android 9.0 (Pie) or API-28 . Earlier releases of aTalk was not fully compatible with API-27 and above. Please feedback to me if you still see problems.

You may refer to http://omemo.top/ and aTalk online https://atalk.sytes.net

**Comment by cmeng-git** [comment]
> @cmeng-git, I just re-tested aTalk and indeed it works a lot better than before! Audio call works, video didn't (camera permission granted but in settings it just says `No camera device found` ). OMEMO messages could have `type="chat"` so they don't pop out in windows in Gajim (non-encrypted messages have that and work OK).
> 
> All in all aTalk looks really good now, I'll definitely keep an eye on it!

wiktor-k: Can you sent the debug log for my investigation on camera issue

type="chat" => Has just fixed in aTalk, but not sure if vanitasvitae should include this fix in Omemo Message instead i.e. `Message sendMessage = encryptedMessage.asMessage(toJid);`; the type is not defined in sendMessage.

By the way can we take the discussion off Jitsi thread - the jitsi team may have some concern.

**Comment by Neustradamus** [comment]
We are in 2020, any news on it?

**Comment by vanitasvitae** [comment]
@Neustradamus we are in 2020, where is your PR?

**Comment by ccoenen** [comment]
*calibrates time machine* shit. wrong year.

**Comment by playforvoices1** [comment]
Would also interested in to see OMEMO support in jitsi. Any updates?

**Comment by smith476** [comment]
my 2 cents:
no updates and probably not occurring in near future: it is both a licence
issue to implement the already existing OMEMO support in the smack library
into jitsi (GPL vs Apache) and of course lack of man power volunteering to
write an OMEMO plugin.
check here my conversion with the jitsi devs about OMEMO on the forum:
https://community.jitsi.org/t/upgrade-to-jitsi-desktop-2-11-5615-on-macos-10-13-wiped-all-xmpp-accounts-new-accounts-not-possible/19545/4

of course I would like to see OMEMO implemented in Jitsi Desktop, imho
usage of XMPP never reached the masses (and will never), currently the
focus using FOSS software in communication (audio/video/chat) is on
Jitsi-Meet/WebRTC, Riot-client/Matrix and Signal.
if Jitsi-meet would implement ZRTP for multiparty videoconferencing (not
sure if technically possible), it would be a killer app (currently WebRTC
is "fully encrypted" but not really  end-to-end but only
client-to-server-to-client; unless you setup your own server and trust it
(...) it is still not a 100% secure solution).
I am using Jitsi Desktop since 10 years, however my XMPP buddies are
vanishing to other clients, only one is left...

Am Mi., 25. März 2020 um 10:55 Uhr schrieb Marko <notifications@github.com>:

**Comment by Neustradamus** [comment]
In the same time: XEP-0384: OMEMO Encryption has been updated to 0.4.0:
- http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.3.0.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.4.0.html

**Comment by damencho** [comment]
@Neustradamus did the license change?

**Comment by lovetox** [comment]
There is no license anymore, the whole crypto approach is now specified, so one can search librarys that support the primitives and implement the crypto themself.

**Comment by Neustradamus** [comment]
@damencho: XEP has always like it, in 2020: "© 1999 – 2020 XMPP Standards Foundation"

**Comment by damencho** [comment]
Sorry @lovetox @Neustradamus, I was talking about https://github.com/igniterealtime/Smack/blob/master/documentation/extensions/omemo.md
`If you want to use OMEMO in a GPLv3 licensed client, you can use the smack-omemo-signal Smack module` that is not compatible with Apache license, but if anybody is willing to implement it under another license, you are welcome to open a PR which will be reviewed.

**Comment by alexanderadam** [comment]
@Neustradamus you are mixing things up here. The XEP-0384 is just the reference. As far as I know you could do an implementation and license this _implementation_ in whatever you want.

_One_ of those _implementations_(!) is [Smack](https://github.com/igniterealtime/Smack). But if you want to use Smack in one of your projects, [it must to be licensed under the GPLv3 or compatible](https://github.com/igniterealtime/Smack/wiki/OMEMO-libsignal-Licensing-Situation).
As [mentioned earlier](https://github.com/jitsi/jitsi/issues/199#issuecomment-315531690) Jitsi Android is licensed under Apache license. This means Jitsi Android cannot use the particular implementation Smack.

**Comment by lovetox** [comment]
The license situation has not changed for smack-omemo-signal, but the XEP now specifies how the crypto works so so its not necessary anymore to depend on code that uses libsignal, someone can now look at the spec and impl the crypto themself in a new library which can be licensed under whatever license the creator wants, after that someone can use that library in the smack plugin and drop smack-omemo-signal

**Comment by vanitasvitae** [comment]
I deliberately designed Smacks OMEMO support with an eventual drop-in replacement of libsignal in mind. While smack-omemo is not yet updated to 0.4.0 (there are some changes needed and all in all the module could use some love), it is totally possible to replace libsignal with another double ratchet library.

**Comment by damencho** [comment]
Nice, good to know. Anyone willing to do it, just ask me questions for little guidance, I can help. But for now, I haven't heard of anybody willing to do the work.

**Comment by Neustradamus** [comment]
For information, after the 0.4.0, there is a 0.5.0.
- https://xmpp.org/extensions/xep-0384.html
- http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.4.0.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.5.0.html
- http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.3.0.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.4.0.html

**Comment by Neustradamus** [comment]
XEP-0384: OMEMO Encryption:
- https://xmpp.org/extensions/xep-0384.html

And XEP-0454: OMEMO Media sharing:
- https://xmpp.org/extensions/xep-0454.html

At this time, it is: 0.9.0 (2025-04-07), previous was 0.8.3 (2022-01-18).

Diff Tool:
- 0.3.0 to current: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.3.0.html&newfile=https://xmpp.org/extensions/xep-0384.html
- 0.8.3 to current: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.8.3.html&newfile=https://xmpp.org/extensions/xep-0384.html
- 0.8.3 to 0.9.0: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.8.3.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.9.0.html
- 0.8.2 to 0.8.3: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.8.2.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.8.3.html
- 0.8.1 to 0.8.2: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.8.1.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.8.2.html
- 0.8.0 to 0.8.1: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.8.0.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.8.1.html
- 0.7.0 to 0.8.0: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.7.0.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.8.0.html
- 0.6.0 to 0.7.0: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.6.0.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.7.0.html
- 0.5.0 to 0.6.0: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.5.0.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.6.0.html
- 0.4.0 to 0.5.0: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.4.0.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.5.0.html
- 0.3.0 to 0.4.0: http://www.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0384-0.3.0.html&newfile=https://xmpp.org/extensions/attic/xep-0384-0.4.0.html
